source/multigrid/mg_transfer_matrix_free.cc

```diff
@@ -355,6 +355,24 @@ MGTransferMatrixFree<dim, Number>::do_prolongate_add(
     Utilities::fixed_power<dim>(n_child_dofs_1d);
   constexpr unsigned int three_to_dim = Utilities::pow(3, dim);
 
+  // If we have user defined MG constraints, we must create
+  // a non-const version of the source vector to distirbute
+  // constraints.
+  LinearAlgebra::distributed::Vector<Number> new_src;
+  {
+    LinearAlgebra::distributed::Vector<Number> copy_src(src);
+
+    // Distribute any user defined constraints
+    this->mg_constrained_dofs->get_user_constraint_matrix(to_level - 1)
+      .distribute(copy_src);
+
+    // Create a new ghosted vector from the copied source vector
+    new_src.reinit(copy_src);
```

Comment:
Can we skip doing this work if no user constraints have been set? This might be difficult to detect as only some of the processors have entries. Maybe only do the .reinit for the constraintmatrix if the user adds constraints? Then you can check the .size() here.

Reply:
Not quite. ``.get_local_lines().size() > 0`` checks if this processor has any constraints and not if custom user constraints are in use. You can check this with ``.size()>0`` with your change to add_user_constraints.
Finally, we still have the problem that you made the code path without constraints slower.
I would do something like:
```
Vector new_src;
Vector* to_use = &src;
if (user_constraints[level].size()>0)
{
// initialize new_src
to_use = &new_src;
}

// use *to_use down here
```